src/main/kotlin/io/perfometer/runner/ThreadPoolScenarioRunner.kt

```diff
@@ -1,66 +1,97 @@
 package io.perfometer.runner
 
 import io.perfometer.dsl.HttpStep
+import io.perfometer.dsl.ParallelStep
 import io.perfometer.dsl.PauseStep
 import io.perfometer.dsl.RequestStep
-import io.perfometer.http.client.HttpClient
 import io.perfometer.http.client.HttpClientFactory
+import io.perfometer.internal.helper.decorateInterruptable
+import io.perfometer.internal.helper.decorateSuspendingInterruptable
 import io.perfometer.statistics.PauseStatistics
 import io.perfometer.statistics.ScenarioSummary
 import kotlinx.coroutines.runBlocking
 import java.time.Duration
-import java.util.concurrent.CompletableFuture
-import java.util.concurrent.Executors
-import java.util.concurrent.TimeUnit
+import java.util.concurrent.*
 
 internal class ThreadPoolScenarioRunner(
     httpClientFactory: HttpClientFactory,
 ) : BaseScenarioRunner(httpClientFactory) {
 
-    private val httpClient = ThreadLocal<HttpClient>()
+    private val httpClient = ThreadLocal.withInitial { httpClientFactory() }
+
+    private val parallelJobs = ThreadLocal.withInitial { ConcurrentLinkedDeque<CompletableFuture<Void>>() }
```

Comment:
You do not need to use Concurrent* collections in ThreadLocal. Thread locals are thread confined

Reply:
Right, this is a leftover.